libraries/AC_Avoidance/AP_OABendyRuler.cpp

```diff
@@ -0,0 +1,294 @@
+        if (i & 1) {
+            // we alternate between trying to the left and right of the target
+            bearing_delta = -bearing_delta;
+        }
+        // bearing that we are probing
+        const float bearing_test = wrap_180(bearing_to_dest + bearing_delta);
+        const float bearing_change_deg = wrap_180(bearing_test - ground_course_deg);
+
+        // don't consider 180 degree turns, as we can't predict which way we will turn
+        if (fabsf(bearing_change_deg) > 170) {
+            continue;
+        }
+
+        // ToDo: add effective groundspeed calculations using airspeed
+        // ToDo: add prediction of vehicle's position change as part of turn to desired heading
+
+        // test location is projected from current location at test bearing
+        Location test_loc = current_loc;
+        test_loc.offset_bearing(bearing_test, lookahead_step1_dist);
+
+        // calculate margin from fence for this scenario
+        float margin = calc_avoidance_margin(current_loc, test_loc);
+        if (margin > best_margin) {
+            best_margin_bearing = bearing_test;
+            best_margin = margin;
+        }
+        if (margin > _margin_max) {
+            // this bearing avoids obstacles out to the lookahead_step1_dist
+            // now check if there is a clear path to the destination
+            if (!have_best_bearing) {
+                best_bearing = bearing_test;
+                have_best_bearing = true;
+            } else if (fabsf(wrap_180(ground_course_deg - bearing_test)) <
+                       fabsf(wrap_180(ground_course_deg - best_bearing))) {
+                // replace bearing with one that is closer to our current ground course
+                best_bearing = bearing_test;
+            }
+
+            // perform second stage test in three directions looking for obstacles
+            const float test_bearings[] = { 0, 45, -45 };
+            const float bearing_to_dest2 = test_loc.get_bearing_to(destination) * 0.01f;
+            float distance2 = constrain_float(lookahead_step2_dist, OA_BENDYRULER_LOOKAHEAD_STEP2_MIN, test_loc.get_distance(destination));
+            for (uint8_t j = 0; j < ARRAY_SIZE(test_bearings); j++) {
+                float bearing_test2 = wrap_180(bearing_to_dest2 + test_bearings[j]);
+                Location test_loc2 = test_loc;
+                test_loc2.offset_bearing(bearing_test2, distance2);
+
+                // calculate minimum margin to fence and obstacles for this scenario
+                float margin2 = calc_avoidance_margin(test_loc, test_loc2);
+                if (margin2 > _margin_max) {
+                    // all good, now project in the chosen direction by the full distance
+                    destination_new = current_loc;
+                    destination_new.offset_bearing(bearing_test, distance_to_dest);
+                    _current_lookahead = MIN(_lookahead, _current_lookahead * 1.1f);
+                    return (i != 0 || j != 0);
```

Comment:
The description on this function is

```
 // returns true and updates origin_new and destination_new if a best path has been found
```

Why is `i == 0 && j == 0` not a valid best path?

Reply:
if i and j are zero then the best path is directly towards the destination so we might as well turn off object avoidance.  I've added a comment to clarify this, "// if the chosen direction is directly towards the destination turn off avoidance"